<a href="https://colab.research.google.com/github/nmuzyka2004/Pandas-BS.../blob/main/2_actividad_numpy_pandas_Natalia_Muzyka.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd

# Numpy

### Ejercicio arrays

1. Crear un array `array_1` de dimensión 6 filas y 4 columnas
2. Crear un array `array_2` cambiando las dimensiones a 3 filas y 8 columnas
3. Crear un array `array_3` cambiando las dimensiones a 5 filas y 7 columnas
4. Crear un array `array_4` de zeros y otro de unos de dimension 2 filas, 3 columnas

In [7]:
# Paso 1: Crear un Array `array_1` de Dimensión 6 Filas y 4 Columnas
array_1 = np.random.randint(0, 100, (6, 4))
print("-------- Array 1 (6x4)-------")
print(array_1)
print('-----------------------------')

# Paso 2: Crear un Array `array_2` Cambiando las Dimensiones a 3 Filas y 8 Columnas
array_2 = array_1.reshape(3, 8)
print("-------- Array 2 (3x8)-------")
print(array_2)
print('-----------------------------')

# Paso 3: Crear un Array `array_3` Cambiando las Dimensiones a 5 Filas y 7 Columnas
array_3 = np.resize(array_1, (5, 7))
print("-------- Array 3 (5x7)-------")
print(array_3)
print('-----------------------------')

# Paso 4: Crear un Array `array_4` de Zeros y Otro de Unos de Dimensión 2 Filas y 3 Columnas
array_zeros = np.zeros((2, 3))
print("----- Array de Zeros (2x3)------")
print(array_zeros)
print('-----------------------------')

array_unos = np.ones((2, 3))
print("----- Array de unos(2x3)-------")
print(array_unos)
print('-----------------------------')


-------- Array 1 (6x4)-------
[[ 2  6 13 94]
 [ 6 91  3 31]
 [73 63  9 93]
 [41 63 61 10]
 [83 79 92 93]
 [13  4 34  0]]
-----------------------------
-------- Array 2 (3x8)-------
[[ 2  6 13 94  6 91  3 31]
 [73 63  9 93 41 63 61 10]
 [83 79 92 93 13  4 34  0]]
-----------------------------
-------- Array 3 (5x7)-------
[[ 2  6 13 94  6 91  3]
 [31 73 63  9 93 41 63]
 [61 10 83 79 92 93 13]
 [ 4 34  0  2  6 13 94]
 [ 6 91  3 31 73 63  9]]
-----------------------------
----- Array de Zeros (2x3)------
[[0. 0. 0.]
 [0. 0. 0.]]
-----------------------------
----- Array de unos(2x3)-------
[[1. 1. 1.]
 [1. 1. 1.]]
-----------------------------


## Ejercicio numpy

Imagina que tienes datos de temperatura promedio diaria de una ciudad durante un mes (30 días). Quieres realizar un análisis básico de estos datos usando Numpy para obtener información útil como la temperatura promedio del mes, la temperatura máxima y mínima, y los días en los que la temperatura estuvo por encima de la temperatura promedio.

1. **Crear un array de temperaturas:**

  - Genera un array de Numpy con 30 elementos que representen las temperaturas promedio diarias de la ciudad durante un mes.
  - Que se genere con numeros enteros aleatorios, entre los valores 18 y 29

2. **Calcular estadísticas básicas:**

  - Calcula la temperatura promedio del mes.
  - Encuentra la temperatura máxima y mínima del mes.

3. **Identificar días calurosos:**

  - Identifica los días en los que la temperatura estuvo por encima de la temperatura promedio del mes.

4. **Imprimir resultados:**

  - Imprime la temperatura promedio, máxima y mínima.
  - Imprime los índices (días) en los que la temperatura estuvo por encima del promedio.

In [8]:
# Paso 1: Crear un array de temperaturas para 30 días
temperaturas = np.random.randint(18,29, 30)
temperaturas



array([26, 18, 24, 20, 22, 26, 27, 25, 24, 27, 24, 22, 26, 22, 23, 26, 24,
       21, 28, 20, 24, 18, 24, 25, 21, 27, 23, 27, 26, 28])

In [10]:
# Paso 2: Calcular estadísticas básicas
temperatura_promedio = np.mean(temperaturas)
temperatura_maxima = np.max(temperaturas)
temperatura_minima = np.min(temperaturas)
print(f"Temperatura promedio del mes: {temperatura_promedio:.2f}")
print(f"Temperatura máxima del mes: {temperatura_maxima}")
print(f"Temperatura mínima del mes: {temperatura_minima}")

Temperatura promedio del mes: 23.93
Temperatura máxima del mes: 28
Temperatura mínima del mes: 18


In [11]:
# Paso 3: Identificar días calurosos
dias_calurosos = np.where(temperaturas > temperatura_promedio)[0]
# np.where para encontrar los índices de los días en que la temperatura estuvo por encima del promedio

# Paso 4: Imprimir resultados
print(f"Días con temperatura por encima del promedio: {dias_calurosos + 1}")  # Añadir 1 para mostrar días del 1 al 30


Días con temperatura por encima del promedio: [ 1  3  6  7  8  9 10 11 13 16 17 19 21 23 24 26 28 29 30]


# Pandas

### Ejercicio pandas

1. Sobre nuestro dataframe "df" crea nuevas columnas:
  - **"age_menos_20":** resultado de restarle 20 años al campo 'edad'
  - **"age_mas_10":** resultado de sumarle 10 años al campo 'edad'
  - **"annual_salary_entre_1000":** resultado de dividir el salario anual entre 1000
  - **"salario_mensual":** resultado de dividir el salario anual entre 12
  - **"edad_respecto_media":** a partir del campo 'Gender', generar otro campo tal que si la edad es mayor que la media se informe con 'Mayor que la media' y si es menor se informe con 'Menor que la media'
  
2. Guarda en una variable de tipo lista los nombres de las columnas y muestra el valor de la variable

3.  Muestra las 20 ultimas filas

4. Elimina las columnas 'age_menos_20' y 'annual_salary_entre_1000' del dataframe 'df'

5. Muestra las 8 primeras filas

6. Crea un nuevo dataframe llamado 'df_2' que no tenga el campo 'age_mas_10' (pero mantenga los campos que teníamos en 'df' hasta el punto 8)

7. Muestra la forma de este nuevo df_2, un resumen de su información y la información estadística de este campo

8. Genera la matriz de correlación y obtén alguna conclusión entre las siguientes variables:
  - Age, AnnualSalary, Purchased

In [24]:
df = pd.read_csv('/content/car_data.csv')
df.head()

User ID Gender  Age  AnnualSalary  Purchased
0      385   Male   35         20000          0
1      681   Male   40         43500          0
2      353   Male   49         74000          0
3      895   Male   40        107500          1
4      661   Male   25         79000          0

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   User ID       1000 non-null   int64 
 1   Gender        1000 non-null   object
 2   Age           1000 non-null   int64 
 3   AnnualSalary  1000 non-null   int64 
 4   Purchased     1000 non-null   int64 
dtypes: int64(4), object(1)
memory usage: 39.2+ KB


In [26]:
# Crea una nueva columna como resultado de restarle 20 años al campo 'edad'
df['age_menos_20'] = df['Age'] - 20

# Crea una nueva columna como resultado de sumarle 10 años al campo 'edad'
df['age_mas_10'] = df['Age'] + 10

# Crea una nueva columna como resultado de dividir el salario anual entre 1000
df["annual_salary_entre_1000"] = df['AnnualSalary'] / 1000

# Crea una nueva columna como resultado de dividir el salario anual entre 12
df["salario_mensual"] = (df['AnnualSalary'] / 12).round(2)

mean_age = df['Age'].mean()
print(f"Media de edades: {mean_age}")

# Crear la nueva columna usando np.where
df['edad_respecto_media'] = np.where(df['Age'] > mean_age, 'Mayor que la media', 'Menor que la media')

df.head()

Media de edades: 40.106


User ID Gender  Age  AnnualSalary  Purchased  age_menos_20  age_mas_10  \
0      385   Male   35         20000          0            15          45   
1      681   Male   40         43500          0            20          50   
2      353   Male   49         74000          0            29          59   
3      895   Male   40        107500          1            20          50   
4      661   Male   25         79000          0             5          35   

   annual_salary_entre_1000  salario_mensual edad_respecto_media  
0                      20.0          1666.67  Menor que la media  
1                      43.5          3625.00  Menor que la media  
2                      74.0          6166.67  Mayor que la media  
3                     107.5          8958.33  Menor que la media  
4                      79.0          6583.33  Menor que la media

In [27]:
# 2. Guarda en una variable de tipo lista los nombres de las columnas y muestra el valor de la variablecolumnad
columnas_df = df.columns.values.tolist()
columnas_df

['User ID',
 'Gender',
 'Age',
 'AnnualSalary',
 'Purchased',
 'age_menos_20',
 'age_mas_10',
 'annual_salary_entre_1000',
 'salario_mensual',
 'edad_respecto_media']

In [28]:
# 3.  Muestra las 20 ultimas filas
df.tail(20)

User ID  Gender  Age  AnnualSalary  Purchased  age_menos_20  age_mas_10  \
980       82  Female   46         23500          1            26          56   
981      695  Female   39         79000          0            19          49   
982      946    Male   39         60500          0            19          49   
983      676    Male   42         54000          0            22          52   
984      649  Female   28         87000          0             8          38   
985      477  Female   45         77500          0            25          55   
986      572    Male   48         33000          1            28          58   
987      832  Female   31         71000          0            11          41   
988      302    Male   33        121500          1            13          43   
989      420  Female   27         84000          0             7          37   
990      971    Male   25         56500          0             5          35   
991      382  Female   30         48500          0            10          40   
992      168  Female   24         67500          0             4          34   
993      476  Female   54        136500          0            34          64   
994      951  Female   53        104000          1            33          63   
995      863    Male   38         59000          0            18          48   
996      800  Female   47         23500          0            27          57   
997      407  Female   28        138500          1             8          38   
998      299  Female   48        134000          1            28          58   
999      687  Female   44         73500          0            24          54   

     annual_salary_entre_1000  salario_mensual edad_respecto_media  
980                      23.5          1958.33  Mayor que la media  
981                      79.0          6583.33  Menor que la media  
982                      60.5          5041.67  Menor que la media  
983                      54.0          4500.00  Mayor que la media  
984                      87.0          7250.00  Menor que la media  
985                      77.5          6458.33  Mayor que la media  
986                      33.0          2750.00  Mayor que la media  
987                      71.0          5916.67  Menor que la media  
988                     121.5         10125.00  Menor que la media  
989                      84.0          7000.00  Menor que la media  
990                      56.5          4708.33  Menor que la media  
991                      48.5          4041.67  Menor que la media  
992                      67.5          5625.00  Menor que la media  
993                     136.5         11375.00  Mayor que la media  
994                     104.0          8666.67  Mayor que la media  
995                      59.0          4916.67  Menor que la media  
996                      23.5          1958.33  Mayor que la media  
997                     138.5         11541.67  Menor que la media  
998                     134.0         11166.67  Mayor que la media  
999                      73.5          6125.00  Mayor que la media

In [29]:
# 4. Elimina las columnas 'age_menos_20' y 'annual_salary_entre_1000' del dataframe 'df'
df.drop(['age_menos_20','annual_salary_entre_1000'], axis=1, inplace=True)
df.head(2)

User ID Gender  Age  AnnualSalary  Purchased  age_mas_10  salario_mensual  \
0      385   Male   35         20000          0          45          1666.67   
1      681   Male   40         43500          0          50          3625.00   

  edad_respecto_media  
0  Menor que la media  
1  Menor que la media

In [30]:
# 5. Muestra las 8 primeras filas
df.head(8)

User ID  Gender  Age  AnnualSalary  Purchased  age_mas_10  salario_mensual  \
0      385    Male   35         20000          0          45          1666.67   
1      681    Male   40         43500          0          50          3625.00   
2      353    Male   49         74000          0          59          6166.67   
3      895    Male   40        107500          1          50          8958.33   
4      661    Male   25         79000          0          35          6583.33   
5      846  Female   47         33500          1          57          2791.67   
6      219  Female   46        132500          1          56         11041.67   
7      588    Male   42         64000          0          52          5333.33   

  edad_respecto_media  
0  Menor que la media  
1  Menor que la media  
2  Mayor que la media  
3  Menor que la media  
4  Menor que la media  
5  Mayor que la media  
6  Mayor que la media  
7  Mayor que la media

In [31]:
# 6. Crea un nuevo dataframe llamado 'df_2' que no tenga el campo 'age_mas_10'
# (pero mantenga los campos que teníamos en 'df' hasta el punto 8)
df_2 = df.drop(['age_mas_10'], axis=1)
df_2.head(2)

User ID Gender  Age  AnnualSalary  Purchased  salario_mensual  \
0      385   Male   35         20000          0          1666.67   
1      681   Male   40         43500          0          3625.00   

  edad_respecto_media  
0  Menor que la media  
1  Menor que la media

In [32]:
# 7. Muestra la forma de este nuevo df_2, un resumen de su información y la información estadística de este campo
df_2.describe()

User ID          Age   AnnualSalary    Purchased  salario_mensual
count  1000.000000  1000.000000    1000.000000  1000.000000      1000.000000
mean    500.500000    40.106000   72689.000000     0.402000      6057.416800
std     288.819436    10.707073   34488.341867     0.490547      2874.028577
min       1.000000    18.000000   15000.000000     0.000000      1250.000000
25%     250.750000    32.000000   46375.000000     0.000000      3864.582500
50%     500.500000    40.000000   72000.000000     0.000000      6000.000000
75%     750.250000    48.000000   90000.000000     1.000000      7500.000000
max    1000.000000    63.000000  152500.000000     1.000000     12708.330000

In [33]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   User ID              1000 non-null   int64  
 1   Gender               1000 non-null   object 
 2   Age                  1000 non-null   int64  
 3   AnnualSalary         1000 non-null   int64  
 4   Purchased            1000 non-null   int64  
 5   salario_mensual      1000 non-null   float64
 6   edad_respecto_media  1000 non-null   object 
dtypes: float64(1), int64(4), object(2)
memory usage: 54.8+ KB


In [35]:
# 8. Genera la matriz de correlación y obtén alguna conclusión entre las siguientes variables
matriz_corr = df[['Age','AnnualSalary','Purchased']].corr()

In [36]:
umbral = 0.4
high_corr_matrix = matriz_corr[(matriz_corr.abs() > umbral) & (matriz_corr != 1)]
high_corr_matrix

Age  AnnualSalary  Purchased
Age                NaN           NaN   0.616036
AnnualSalary       NaN           NaN        NaN
Purchased     0.616036           NaN        NaN

### Ejercicio rename

1. Cambia el nombre de las siguientes columnas sobre el df 'df_renamed'

- Age -> edad
- AnnualSalary -> salario_anual
- Purchased -> comprado

In [37]:
# Age -> edad
df.rename(columns = {"Age": "edad"}, inplace = True)
# AnnualSalary -> salario_anual
df.rename(columns = {"AnnualSalary": "salario_anual"}, inplace = True)
# Purchased -> comprado
df.rename(columns = {"Purchased": "comprado"}, inplace = True)

df.head()

User ID Gender  edad  salario_anual  comprado  age_mas_10  salario_mensual  \
0      385   Male    35          20000         0          45          1666.67   
1      681   Male    40          43500         0          50          3625.00   
2      353   Male    49          74000         0          59          6166.67   
3      895   Male    40         107500         1          50          8958.33   
4      661   Male    25          79000         0          35          6583.33   

  edad_respecto_media  
0  Menor que la media  
1  Menor que la media  
2  Mayor que la media  
3  Menor que la media  
4  Menor que la media

### Ejercicio filter

1. df_filtrado_1: edad mayor que 30 años, salario menor a 100.000 y género ha de ser female

2. df_filtrado_2: salario menor que 100.000, edad mayor o igual a 45 y ordenar por su edad de mayor a menor

3. df_filtrado_3: que hayan realizado alguna compra, que sean 'male', que tengan más de 30 años pero menos de 50 y que su salario sea mayor a 100.000, ordenados por edad (de menor a mayor) y salario (de mayor a menor)

In [39]:
df = pd.read_csv('/content/car_data.csv')
df.head()

User ID Gender  Age  AnnualSalary  Purchased
0      385   Male   35         20000          0
1      681   Male   40         43500          0
2      353   Male   49         74000          0
3      895   Male   40        107500          1
4      661   Male   25         79000          0

In [50]:
# 1. Edad mayor que 30 años, salario menor a 100.000 y género ha de ser female
df_filtrado_1 = df[(df["Age"] > 30) & (df["AnnualSalary"] < 100000) & (df["Gender"] == "Female")]
# comprobamos que se haya filtrado correctamente
print("\n>>> El unico valor en genero ha de ser Female: ")
print(df_filtrado_1.Gender.unique())

print("\n>>> El salario anual maximo ha de ser menor a 100.000: ")
print(df_filtrado_1.AnnualSalary.max())

print("\n>>> La edad minima ha de ser 30 años: ")
print(df_filtrado_1.Age.min())

print("\n>>> El df_filtrado_1:")
df_filtrado_1.head()


>>> El unico valor en genero ha de ser Female: 
['Female']

>>> El salario anual maximo ha de ser menor a 100.000: 
99500

>>> La edad minima ha de ser 30 años: 
31

>>> El df_filtrado_1:


User ID  Gender  Age  AnnualSalary  Purchased
5       846  Female   47         33500          1
12      790  Female   32         72500          0
20      338  Female   39         51500          0
21      134  Female   49         39000          1
23      294  Female   41         61500          0

In [51]:
# 2. Salario menor que 100.000, edad mayor o igual a 45 y ordenar por su edad de mayor a menor
df_filtrado_2 = df[ (df["Age"] >= 45) & (df["AnnualSalary"] < 100000)].sort_values(by="Age", ascending=False)

# comprobamos que se haya filtrado correctamente
print("\n>>> El salario anual maximo ha de ser menor a 100.000: ")
print(df_filtrado_2.AnnualSalary.max())

print("\n>>> La edad minima ha de ser 45 años: ")
print(df_filtrado_2.Age.min())

print("\n>>> El df_filtrado_2:")
df_filtrado_2.head()


>>> El salario anual maximo ha de ser menor a 100.000: 
98500

>>> La edad minima ha de ser 45 años: 
45

>>> El df_filtrado_2:


User ID  Gender  Age  AnnualSalary  Purchased
168      443  Female   63         48500          1
972      440    Male   63         85500          1
622       18    Male   63         44500          1
424      880    Male   63         36500          1
787      431    Male   63         44500          1

In [52]:
# 3. Que hayan realizado alguna compra, que sean 'male', que tengan más de 30 años pero menos de 50 y que su salario sea mayor a 100.000
# ordenados por edad y salario
df_filtrado_3 = df[ (df["Purchased"]==1) & (df["Gender"]=="Male") & (df["Age"] > 30) & (df["Age"]<50) & (df["AnnualSalary"] > 100000)] \
                .sort_values(by=["Age","AnnualSalary"], ascending=[True, False])

# comprobamos que se haya filtrado correctamente
print("\n>>> El unico valor en purchased ha de ser 1: ")
print(df_filtrado_3.Purchased.unique())

print("\n>>> El unico valor en genero ha de ser Male: ")
print(df_filtrado_3.Gender.unique())

print("\n>>> El salario anual mínimo ha de ser mayor a 100.000: ")
print(df_filtrado_3.AnnualSalary.min())

print("\n>>> La edad minima ha de ser 30 años: ")
print(df_filtrado_3.Age.min())

print("\n>>> La edad máxima ha de ser 50 años: ")
print(df_filtrado_3.Age.max())

print("\n>>> El df_filtrado_3:")
df_filtrado_3.head()


>>> El unico valor en purchased ha de ser 1: 
[1]

>>> El unico valor en genero ha de ser Male: 
['Male']

>>> El salario anual mínimo ha de ser mayor a 100.000: 
100500

>>> La edad minima ha de ser 30 años: 
31

>>> La edad máxima ha de ser 50 años: 
49

>>> El df_filtrado_3:


User ID Gender  Age  AnnualSalary  Purchased
164       42   Male   31        136500          1
882      451   Male   31        108500          1
463      685   Male   32        120000          1
988      302   Male   33        121500          1
716      955   Male   33        101500          1

### Ejercicio group by

1. Agrupar por `Product` y `Region`, y obtener la suma de las ventas
2. Filtrar el df para incluir solo aquellos productos donde las ventas son mayores que 1000, agrupar por `Product`, y obtener el promedio de las ventas para cada producto
3. Agrupar por `Region`, y obtener la suma de las ventas y el número total de ventas
4. Filtrar el df para incluir solo aquellas regiones donde la suma de ventas es mayor que 2000, agrupar por `Region`, y obtener la suma y el promedio de las ventas para cada región (usar lambda!!!)

In [122]:
# Crear el df de ejemplo
data = {
    'Product': ['A', 'B', 'A', 'C', 'B', 'C', 'A', 'B', 'C', 'A'],
    'Region': ['North', 'South', 'East', 'West', 'North', 'South', 'East', 'West', 'North', 'South'],
    'Sales': [1000, 1500, 800, 1200, 900, 1100, 750, 1300, 850, 950]
}

df = pd.DataFrame(data)
df

Product Region  Sales
0       A  North   1000
1       B  South   1500
2       A   East    800
3       C   West   1200
4       B  North    900
5       C  South   1100
6       A   East    750
7       B   West   1300
8       C  North    850
9       A  South    950

In [123]:
# 1: Agrupar por `Product` y `Region`, y obtener la suma de las ventas.
sales_sum_product_region = df.groupby(['Product', 'Region'])['Sales'].sum().reset_index()
sales_sum_product_region


Product Region  Sales
0       A   East   1550
1       A  North   1000
2       A  South    950
3       B  North    900
4       B  South   1500
5       B   West   1300
6       C  North    850
7       C  South   1100
8       C   West   1200

In [124]:
# 2: Filtrar el df para incluir solo aquellos productos donde las ventas son mayores que 1000,
# agrupar por `Product`, y obtener el promedio de las ventas para cada producto.
filtered_df = df[df['Sales'] > 1000]  # Filtrar para ventas mayores que 1000
filtered_df

Product Region  Sales
1       B  South   1500
3       C   West   1200
5       C  South   1100
7       B   West   1300

In [125]:
average_sales_product = filtered_df.groupby('Product')['Sales'].mean().reset_index()  # Calcular promedio de ventas por producto
average_sales_product

Product   Sales
0       B  1400.0
1       C  1150.0

In [126]:
# 3: Agrupar por `Region`, y obtener la suma de las ventas y el número total de ventas.
total_sales_region = df.groupby('Region')['Sales'].agg(['sum', 'count']).reset_index()  # Calcular suma y contar ventas por región
total_sales_region

Region   sum  count
0   East  1550      2
1  North  2750      3
2  South  3550      3
3   West  2500      2

In [127]:
# 4: Filtrar el df para incluir solo aquellas regiones donde la suma de ventas es mayor que 2000,
# agrupar por `Region`, y obtener la suma y el promedio de las ventas para cada región.
# Filtrar regiones con suma de ventas mayor que 2000
filtered_df = df.groupby('Region').filter(lambda x: x['Sales'].sum() > 2000)
filtered_df.sort_values(by="Region")

# Calcular suma y promedio de ventas por región
sales_stats_region = filtered_df.groupby('Region')['Sales'].agg(['sum', 'mean']).round(2).reset_index()
sales_stats_region

Region   sum     mean
0  North  2750   916.67
1  South  3550  1183.33
2   West  2500  1250.00

### Ejercicio pivot & crosstab (opcional)

Tenemos un df que contiene información sobre las compras realizadas por los usuarios en una tienda en línea. Las columnas son las siguientes:

- **User ID:** Identificador único del usuario
- **Gender:** Género del usuario (Male/Female)
- **Age:** Edad del usuario
- **Product Category:** Categoría del producto comprado (Electronics, Clothing, Home, etc.)
- **Purchase Amount:** Monto de la compra

Tenemos que:

1. Crear una tabla dinámica que muestre el monto *total* de las compras por edad y categoría de producto
  - ¿Qué categorías de productos son las más compradas en cada grupo de edad?

2. Crear una tabla de contingencia que muestre la frecuencia de compras por Gender y Product Category.

In [ ]:
import pandas as pd
import numpy as np

# Crear el DataFrame de ejemplo
data = {
    'User ID': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    'Gender': ['Male', 'Female', 'Female', 'Male', 'Female', 'Male', 'Male', 'Female', 'Male', 'Female'],
    'Age': [23, 25, 31, 23, 40, 27, 30, 27, 40, 24],
    'Product Category': ['Electronics', 'Clothing', 'Clothing', 'Home', 'Electronics', 'Home', 'Clothing', 'Home', 'Electronics', 'Clothing'],
    'Purchase Amount': [150, 200, 100, 50, 300, 75, 60, 45, 400, 80]
}

df = pd.DataFrame(data)
df


In [ ]:
# Crear la tabla dinámica


In [ ]:
# Encontrar la categoría de productos más comprada en cada grupo de edad


In [ ]:
# Crear la tabla de contingencia


In [ ]:
# Encontrar la categoría de productos más popular entre hombres y mujeres


# Ejercicio Streaming

1. Eliminar las columnas con información de id de sesiones, id de dispositivos, lenguaje, localización, calidad de playback y cantidad de interacciones
2. Genero preferido por los hombres
3. Corroborar si los smartphones son el dispositivo en el cual permanecen más tiempo los usuarios mirando streaming
4. Las edades que suelen pagar más Premium
5. El video que tiene más tiempo de reproducción
6. El género que más le gustó a las mujeres (rating)


Una vez realizados los apartados, redactar conclusiones que podríamos obtener de los datos. Como analistas, es importante que los datos sepamos entenderlos y sacar insights de los mismos. Podemos crear hipótesis si no tenemos evidencia suficiente para reflejar X cosa como hecho verídico.

Recurso: streaming.csv

In [129]:
import pandas as pd

import io
# Subimos el archivo a nuestro proyecto
from google.colab import files
uploaded = files.upload()


Saving streaming (3).csv to streaming (3).csv


In [137]:
filename = list(uploaded.keys())[0]
df = pd.read_csv(io.BytesIO(uploaded[filename]))
df.head()

User_ID                            Session_ID  \
0  eb4f9229-74df-45f6-baac-cf19241b8b30  cb2142a7-0750-49ed-b8ac-a975fe1ff69a   
1  661d4b59-4328-410a-901c-1e3b4c40c334  3bc0a662-b353-4015-8b0c-55ceb510d13a   
2  dd3fe9e9-ea82-4891-ab93-8a47c80e3251  bd545b4a-9f54-4e87-b9f8-15ae20b44f22   
3  a1b3365b-1d00-4ddf-bc43-02fc9c10c680  0441086d-c59e-478d-a496-5c5b995ecfdb   
4  338d3f91-5f1c-4590-8803-324901826406  0295f01d-7f15-4799-856c-90c688697ef8   

   Device_ID  Video_ID  Duration_Watched (minutes)        Genre  \
0        232        11                   90.044525       Sci-Fi   
1        549        85                   68.973479       Comedy   
2        844        50                   42.511343       Comedy   
3        201        38                   53.316660  Documentary   
4        700        31                   69.437786       Action   

                            Country  Age  Gender Subscription_Status  Ratings  \
0                             Sudan   56  Female             Premium        3   
1                              Cuba   52    Male             Premium        3   
2                             Japan   14  Female             Premium        3   
3  Lao People's Democratic Republic   36    Male             Premium        2   
4                        Bangladesh   31    Male             Premium        4   

  Languages Device_Type          Location Playback_Quality  Interaction_Events  
0   Spanish  Smartphone         Reedshire               4K                  73  
1   Chinese     Desktop       Stevenhaven               SD                  22  
2   Spanish      Tablet        Vaughntown               HD                  41  
3   Spanish      Laptop  East Raymondbury               SD                  40  
4    German      Laptop    Michaelchester               HD                  41

In [138]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6214 entries, 0 to 6213
Data columns (total 16 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   User_ID                     6214 non-null   object 
 1   Session_ID                  6214 non-null   object 
 2   Device_ID                   6214 non-null   int64  
 3   Video_ID                    6214 non-null   int64  
 4   Duration_Watched (minutes)  6214 non-null   float64
 5   Genre                       6214 non-null   object 
 6   Country                     6214 non-null   object 
 7   Age                         6214 non-null   int64  
 8   Gender                      6214 non-null   object 
 9   Subscription_Status         6214 non-null   object 
 10  Ratings                     6214 non-null   int64  
 11  Languages                   6214 non-null   object 
 12  Device_Type                 6214 non-null   object 
 13  Location                    6214 

In [139]:
df.columns

Index(['User_ID', 'Session_ID', 'Device_ID', 'Video_ID',
       'Duration_Watched (minutes)', 'Genre', 'Country', 'Age', 'Gender',
       'Subscription_Status', 'Ratings', 'Languages', 'Device_Type',
       'Location', 'Playback_Quality', 'Interaction_Events'],
      dtype='object')

In [140]:
# Eliminar las columnas con información de id de sesiones, id de dispositivos, lenguaje, localización, calidad de playback y cantidad de interacciones
df.drop(['Session_ID','Device_ID','Languages','Location','Playback_Quality','Interaction_Events'], axis=1, inplace=True)
df.head()


User_ID  Video_ID  Duration_Watched (minutes)  \
0  eb4f9229-74df-45f6-baac-cf19241b8b30        11                   90.044525   
1  661d4b59-4328-410a-901c-1e3b4c40c334        85                   68.973479   
2  dd3fe9e9-ea82-4891-ab93-8a47c80e3251        50                   42.511343   
3  a1b3365b-1d00-4ddf-bc43-02fc9c10c680        38                   53.316660   
4  338d3f91-5f1c-4590-8803-324901826406        31                   69.437786   

         Genre                           Country  Age  Gender  \
0       Sci-Fi                             Sudan   56  Female   
1       Comedy                              Cuba   52    Male   
2       Comedy                             Japan   14  Female   
3  Documentary  Lao People's Democratic Republic   36    Male   
4       Action                        Bangladesh   31    Male   

  Subscription_Status  Ratings Device_Type  
0             Premium        3  Smartphone  
1             Premium        3     Desktop  
2             Premium        3      Tablet  
3             Premium        2      Laptop  
4             Premium        4      Laptop

In [141]:
#  Genero preferido por los hombres
genero_preferido = df[df['Gender'] == 'Male']['Genre'].value_counts().idxmax()
print(f"El genero que prefieren los hombres: {genero_preferido}")

El genero que prefieren los hombres: Documentary


In [147]:
# Corroborar si los smartphones son el dispositivo en el cual permanecen más tiempo los usuarios mirando streaming
tiempo_dispositivo = df.groupby('Device_Type')['Duration_Watched (minutes)'].sum()
tiempo_dispositivo

Device_Type
Desktop       76925.913710
Laptop        72492.588158
Smart TV      78276.340378
Smartphone    75226.388700
Tablet        72270.075235
Name: Duration_Watched (minutes), dtype: float64

In [149]:
dispositivo_max = tiempo_dispositivo.idxmax()
tiempo_max = round(tiempo_dispositivo.max(), 2)
print(f"El dispositivo en el cual permanecen más tiempo los usuarios mirando streaming: {dispositivo_max} ({tiempo_max} minutes)")

El dispositivo en el cual permanecen más tiempo los usuarios mirando streaming: Smart TV (78276.34 minutes)


In [150]:
# Las edades que suelen pagar más Premium
usuarios_Premium = df[df['Subscription_Status'] == 'Premium']
edad_counts = usuarios_Premium['Age'].value_counts().sort_values(ascending=False).reset_index()
# Тор-10 de edad
edad_counts.head(10)

Age  count
0   40     72
1   13     69
2   52     66
3   57     64
4   50     63
5   16     60
6   45     59
7   20     59
8   51     58
9   27     57

In [151]:
# El video que tiene más tiempo de reproducción
max_tiempo_video = df.groupby('Video_ID')['Duration_Watched (minutes)'].sum().round(2)
max_tiempo_video.head()

Video_ID
1    2824.29
2    3677.83
3    3788.64
4    4210.02
5    3949.46
Name: Duration_Watched (minutes), dtype: float64

In [152]:
video_ID_max = max_tiempo_video.idxmax()
tiempo_max = max_tiempo_video.max()
print(f"El video que tiene más tiempo de reproducción: con ID {video_ID_max} ({tiempo_max} minutes)")

El video que tiene más tiempo de reproducción: con ID 11 (5088.84 minutes)


In [153]:
# El género que más le gustó a las mujeres (rating)
female_genre_rating = df[df['Gender'] == 'Female'].groupby('Genre')['Ratings'].mean().round(2)
female_genre_rating

Genre
Action         3.02
Comedy         3.10
Documentary    2.85
Drama          3.01
Sci-Fi         3.06
Thriller       2.94
Name: Ratings, dtype: float64

In [118]:
female_genre_max = female_genre_rating.idxmax()
female_rating_max = female_genre_rating.max()
print(f"El género que más le gustó a las mujeres: {female_genre_max} con rating {female_rating_max}")

El género que más le gustó a las mujeres: Comedy con rating 3.1


1.Según el análisis de los datos proporcionados, la mayoría de los hombres en la muestra prefieren las películas documentales. Por lo general, las preferencias por un género u otro se forman a partir de múltiples factores, siendo los más comunes la edad, el nivel educativo, las características personales, los intereses, el entorno, etc.
Es posible que en esta muestra la mayoría de los hombres tengan edad de más de 35 años o(y) cuenten con estudios superiores o(y) posean una mentalidad técnica o analítica.
Con los datos disponibles, se puede determinar a qué grupo de edad pertenece la mayoría de los hombres en la muestra. Sin embargo, no hay información sobre otros factores influyentes, por lo que se requieren estudios adicionales.

In [168]:
# Filtramos los hombres de tres grupos
df_menos_35_hombres = df[ (df["Age"] < 35) & (df["Gender"] == 'Male')].sort_values(by="Age", ascending=False)
df_menos_35_hombres.head()

User_ID  Video_ID  \
6211  3daafeb1-4075-4aca-a7ce-e3506b7011d7        22   
262   65bdab19-3732-421e-8e7b-e864313d97b1        36   
5902  3f216b3a-f87e-4b09-b108-88f02132e751        81   
6054  e789b5b8-7639-4cbf-8294-d20193808c21        80   
269   05abd77e-c519-4549-8d6d-904c78da31c7        86   

      Duration_Watched (minutes)   Genre                           Country  \
6211                  105.516389   Drama                      Saint Martin   
262                    28.380573   Drama                      Saint Martin   
5902                   91.872923   Drama                             Korea   
6054                   59.036504  Sci-Fi                            Rwanda   
269                    82.008057   Drama  Saint Vincent and the Grenadines   

      Age Gender Subscription_Status  Ratings Device_Type  
6211   34   Male             Premium        5      Tablet  
262    34   Male             Premium        5  Smartphone  
5902   34   Male                Free        3     Desktop  
6054   34   Male             Premium        4  Smartphone  
269    34   Male                Free        2     Desktop

In [169]:
df_entre_35_60_hombres = df[ (df["Age"] >= 35) & (df["Age"] <= 60) & (df["Gender"] == 'Male')].sort_values(by="Age", ascending=False)
df_entre_35_60_hombres.head()

User_ID  Video_ID  \
26    20c0b7a3-2051-4b7d-9c42-fd924069bbc6        13   
1698  62849702-99b9-463f-b7b5-1c19f3d3444d        92   
1758  fa470030-8659-460d-99f1-a8e10db8acd4        87   
1486  8d1dcebd-feb1-4c62-a284-2f75a2a37717        43   
222   84d0de0c-0755-444f-ade7-c39aefae73a7        36   

      Duration_Watched (minutes)        Genre    Country  Age Gender  \
26                     70.180555  Documentary    Lesotho   60   Male   
1698                   66.734588       Sci-Fi    Moldova   60   Male   
1758                   13.468592        Drama  Swaziland   60   Male   
1486                  109.497930     Thriller      Qatar   60   Male   
222                    51.333108       Comedy    Bolivia   60   Male   

     Subscription_Status  Ratings Device_Type  
26               Premium        3      Laptop  
1698             Premium        3    Smart TV  
1758             Premium        5     Desktop  
1486                Free        4     Desktop  
222                 Free        2    Smart TV

In [167]:
df_mas_60_hombres = df[ (df["Age"] > 60) & (df["Gender"] == 'Male')].sort_values(by="Age", ascending=True)
df_mas_60_hombres.head()


User_ID  Video_ID  \
5015  eaec3f32-35ed-4a4d-8b38-49b93032baab        25   
4993  491f8a6e-3eee-4444-9f31-17a7b3fea386        12   
4214  3e573016-6538-4607-9365-64811f16ac84        86   
367   9d99cc4e-232b-4c9a-ab8d-0f7dcc85f477        90   
4292  3f40e377-d07d-4755-8c46-fa99a2f88e2e        70   

      Duration_Watched (minutes)     Genre                Country  Age Gender  \
5015                   29.757280     Drama     Russian Federation   61   Male   
4993                  119.669923    Action  Palestinian Territory   61   Male   
4214                    7.834081  Thriller            El Salvador   61   Male   
367                    10.509861    Comedy               Portugal   61   Male   
4292                   73.671435     Drama          Faroe Islands   61   Male   

     Subscription_Status  Ratings Device_Type  
5015                Free        2  Smartphone  
4993             Premium        4  Smartphone  
4214             Premium        3      Tablet  
367                 Free        1     Desktop  
4292             Premium        2      Laptop

In [170]:
print(f"Cantidad de los hombres de edad menos de 35: {df_menos_35_hombres.shape[0]}")
print(f"Cantidad de los hombres de edad entre 35 y 60: {df_entre_35_60_hombres.shape[0]}")
print(f"Cantidad de los hombres de edad 60+: {df_mas_60_hombres.shape[0]}")

Cantidad de los hombres de edad menos de 35: 1253
Cantidad de los hombres de edad entre 35 y 60: 1341
Cantidad de los hombres de edad 60+: 472


Podemos ver que, efectivamente, la mayoría de los hombres en la muestra tienen más de 35 años. Y esto explica los resultados obtenidos anteriormente sobre las preferencias de los hombres en nuestra muestra. Por lo general, los jóvenes suelen elegir acción, ciencia ficción, comedias y thrillers, ya que buscan dinamismo, espectáculo y adrenalina. Los hombres de mediana edad (35-60 años) pueden estar más interesados en películas históricas, bélicas y documentales, ya que contienen hechos, análisis y un significado profundo. La generación mayor (60+) prefiere el cine clásico, los documentales y las biografías, ya que valoran la historia, el patrimonio cultural y el trasfondo significativo.

2. Según los estudios la mayoría de los encuestados prefieren ver películas utilizando Smart TV. La suposición de que la mayoría veía películas en un smartphone resultó ser errónea.

Es posible que la mayoría de los usuarios en la muestra sean familias o personas mayores de 30 años. En este caso, Smart TV es una elección natural para disfrutar de una experiencia cómoda en casa con una pantalla grande.

Los usuarios más jóvenes y sin familia, posiblemente, usen más los smartphones o tablets para ver contenido.









In [175]:
df_menos_30 = df[(df["Age"] < 30)]
df_menos_30.head()

User_ID  Video_ID  \
2   dd3fe9e9-ea82-4891-ab93-8a47c80e3251        50   
6   91d4aa91-da98-4ced-851d-c5c0194c63c4        85   
7   e707ea19-8d7e-476e-9bf8-c379f01f79da        63   
17  58cc2409-e8d8-4221-b313-125b277741aa         8   
20  d1c113dc-cee2-4c56-ae62-ab274b4fd866        39   

    Duration_Watched (minutes)   Genre              Country  Age  Gender  \
2                    42.511343  Comedy                Japan   14  Female   
6                    63.979099  Action               Guyana   10  Female   
7                    78.338366  Comedy               Guinea   28  Female   
17                   47.919761  Sci-Fi              Estonia   27  Female   
20                  101.476776  Action  Trinidad and Tobago   24  Female   

   Subscription_Status  Ratings Device_Type  
2              Premium        3      Tablet  
6                 Free        2      Laptop  
7                 Free        1     Desktop  
17             Premium        2  Smartphone  
20                Free        2  Smartphone

In [176]:
dispositivo_edad_menos_30 = df_menos_30.groupby('Device_Type')['Duration_Watched (minutes)'].sum()
dispositivo_edad_menos_30

Device_Type
Desktop       26743.457513
Laptop        25101.952408
Smart TV      25058.520149
Smartphone    22621.112455
Tablet        26322.605314
Name: Duration_Watched (minutes), dtype: float64

Podemos ver que la mayoria de los jovenes de edad menos de 30 usan más Desktop o la Tablet que otros dispositivos

In [178]:
df_mas_30 = df[(df["Age"] >= 30)]
df_mas_30.head()

User_ID  Video_ID  Duration_Watched (minutes)  \
0  eb4f9229-74df-45f6-baac-cf19241b8b30        11                   90.044525   
1  661d4b59-4328-410a-901c-1e3b4c40c334        85                   68.973479   
3  a1b3365b-1d00-4ddf-bc43-02fc9c10c680        38                   53.316660   
4  338d3f91-5f1c-4590-8803-324901826406        31                   69.437786   
5  076b03f1-fd5b-4abe-b22e-cf3b4fdad381        83                   41.810912   

         Genre                           Country  Age  Gender  \
0       Sci-Fi                             Sudan   56  Female   
1       Comedy                              Cuba   52    Male   
3  Documentary  Lao People's Democratic Republic   36    Male   
4       Action                        Bangladesh   31    Male   
5       Comedy                           Tokelau   53    Male   

  Subscription_Status  Ratings Device_Type  
0             Premium        3  Smartphone  
1             Premium        3     Desktop  
3             Premium        2      Laptop  
4             Premium        4      Laptop  
5             Premium        4      Laptop

In [179]:
dispositivo_edad_mas_30 = df_mas_30.groupby('Device_Type')['Duration_Watched (minutes)'].sum()
dispositivo_edad_mas_30

Device_Type
Desktop       50182.456197
Laptop        47390.635749
Smart TV      53217.820229
Smartphone    52605.276245
Tablet        45947.469921
Name: Duration_Watched (minutes), dtype: float64

La mayoria de las personas de edad 30+ usan más Smart TV que otros dispositivos

3. En cuanto a la edad de aquellos que compran la suscripción Premium, no se ha identificado una relación clara. En realidad, esto depende más de factores como el nivel económico, la disponibilidad de tiempo libre, las características culturales y la preferencia del cine sobre otras formas de entretenimiento. Para determinar con mayor precisión la influencia de estos factores, se requieren investigaciones adicionales y más datos.

In [190]:
# Generamos la matriz de correlación
matriz_corr = edad_counts.corr()
matriz_corr

Age     count
Age    1.000000 -0.075171
count -0.075171  1.000000

4. El video que tiene más tiempo de reproducción es el que tiene el identificador 11. Podemos ver mas datos sobre este video.



In [185]:
df_video_id_11 = df[(df["Video_ID"] == 11)]
df_video_id_11.head(10)

User_ID  Video_ID  \
0     eb4f9229-74df-45f6-baac-cf19241b8b30        11   
106   5cb241d4-7247-4539-a05c-660f0fcde4f6        11   
221   7ee621fa-9177-4bee-9f73-aa60d3b3b807        11   
281   b6b20872-df06-4ecb-a782-687bb03c3ca3        11   
313   6703d5d0-6fa1-4213-a5b5-80702f608edd        11   
837   b65f2559-ce59-4619-916e-fa8925402e50        11   
937   c27a52a9-2042-414b-9852-4123dd3c4940        11   
1283  3a11f140-8a42-41ab-9b0c-ebd989f12309        11   
1539  2abe3cf3-64ef-4c8d-b9fc-c4d169c5be46        11   
1690  9e3e55c1-15de-4c70-a361-8151167ef8cc        11   

      Duration_Watched (minutes)     Genre                    Country  Age  \
0                      90.044525    Sci-Fi                      Sudan   56   
106                    65.510480     Drama                Isle of Man   55   
221                    64.196802  Thriller           Papua New Guinea   31   
281                    65.686458    Comedy                      Congo   39   
313                   103.677152  Thriller              Cote d'Ivoire   25   
837                    62.246027  Thriller                    Vanuatu   45   
937                    52.682262    Sci-Fi                   Kiribati   50   
1283                   74.649525    Action                      Nepal   36   
1539                   29.584976    Sci-Fi  Saint Pierre and Miquelon   58   
1690                   65.268290     Drama                  Australia   30   

      Gender Subscription_Status  Ratings Device_Type  
0     Female             Premium        3  Smartphone  
106     Male                Free        4     Desktop  
221   Female                Free        1     Desktop  
281   Female                Free        4      Tablet  
313   Female                Free        5  Smartphone  
837     Male             Premium        5      Tablet  
937     Male             Premium        4     Desktop  
1283  Female                Free        2     Desktop  
1539    Male                Free        4      Laptop  
1690    Male             Premium        2  Smartphone

In [184]:
df_video_id_11.shape

(77, 10)

In [188]:
video_11_rating = df_video_id_11['Ratings'].mean().round(2)
print(video_11_rating)

3.25


El video con ID 11 tiene el rating 3.25. En total, este ID aparece 77 veces en la tabla. Sin embargo, se puede observar que bajo este ID hay películas de diferentes géneros. Lo más probable es que no se trate de una sola película. Puede ser un error en el registro de datos o un grupo de películas de un mismo autor o estudio cinematográfico. Para confirmar esto, se necesitan datos adicionales.

5. El género que más les gustó a las mujeres fue Comedy, con un rating de 3.1. Esto puede indicar que la audiencia femenina valora un contenido ligero y relajante, que mejora el estado de ánimo y ayuda a desconectarse de las preocupaciones diarias.








